Do not use '.NET Core' anymore in error messages

vitek-karas opened the pull request:
Added a test for a single-file FDD running over old hostfxr

In reaction to https://github.com/dotnet/runtime/discussions/67427.

vitek-karas committed:
Do not use '.NET Core' anymore in error messages

Added a test for a single-file FDD running over old hostfxr

vitek-karas committed:
Improve the error message

Include apphost version, .NET location and so on. Similar to other improved error messages now.

vitek-karas commented:
@elinor-fung  could you please take another look? I followed the same format as the other "improved" error messages.

elinor-fung commented on src/native/corehost/corehost.cpp:
```diff
@@ -84,12 +84,29 @@ bool is_exe_enabled_for_execution(pal::string_t* app_dll)
+void need_newer_framework_error(const pal::string_t& dotnet_root)
 {
-    pal::string_t url = get_download_url();
-    trace::error(_X("  _ To run this application, you need to install a newer version of .NET Core."));
```
Can we get rid of the check for the `  _ ` prefix now?
https://github.com/dotnet/runtime/blob/febeba309d6606f2c3ca0ee5ef4ba161596e8a25/src/native/corehost/apphost/apphost.windows.cpp#L116

vitek-karas replied:
Isn't it necessary for the pre 7.0 runtimes?
7.0 doesn't use it anymore, but the apphost still needs to work downlevel for some rather weird cases.

elinor-fung replied:
I thought apphost itself was the only thing actually writing with that prefix? Because with downlevel hostfxr, we couldn't actually redirect the errors from hostfxr.

vitek-karas replied:
What about 7.0 `apphost` running against 6.0 `hostfxr` - the `hostfxr` might still use the `_` prefix somewhere (I didn't check). In that case the error info will be redirected from `hostfxr` to `apphost` and if it contains `_` it should be parsed.

elinor-fung replied:
I was pretty sure we only ever used it for `apphost`, but I didn't actually go back and check.

Either way, the UX and changes you have right now look good to me.

vitek-karas replied:
I just searched 3.1 and it only uses the prefix in apphost - but I'm still nervous about doing this. Honestly I would rather keep it.

elinor-fung commented on src/native/corehost/corehost.cpp:
```diff
@@ -84,12 +84,29 @@ bool is_exe_enabled_for_execution(pal::string_t* app_dll)
-    trace::error(_X(""));
-    trace::error(_X("  - %s&apphost_version=%s"), url.c_str(), _STRINGIFY(COMMON_HOST_PKG_VER));
+    pal::string_t host_path;
+    pal::get_own_executable_path(&host_path);
```
Pass in `host_path`? We should already have it in places that use this, so we can avoid getting it again.

elinor-fung commented on src/native/corehost/corehost.cpp:
```diff
@@ -238,7 +255,7 @@ int exe_start(const int argc, const pal::char_t* argv[])
             // This check exists to provide an error message for UI apps when running 3.0 apps on 2.0 only hostfxr, which doesn't support error writer redirection. 
```
Do we actually need to do this anymore now that we are not relying on a custom prefix to show a different message?

elinor-fung replied:
Oh, I guess we are only using it for the URL now? Maybe the dialog should just default to `get_download_url()` + `apphost_version=...`  if it doesn't parse one out of the buffered error messages.

vitek-karas replied:
It's actually also used to write something meaningful to eventlog. On Windows we always redirect error writer. If it's a console app, it's only used to write the error to eventlog. If it's a GUI app we also show the dialog.

Without this, we would have to add similar handling to what the dialog code does to figure out reasonable error message.

Technically the error code is sufficient, but I think it's better to have a nicer error even in the event log (for example it will now contain the apphost version, architecture, ...).

vitek-karas replied:
I updated the comment to make this clearer.

elinor-fung reviewed (approved):
Looks good to me.
Minor comment/question about possible cleanup around the dialog for GUI apps.

vitek-karas committed:
Improve dialog for the case of old runtime

It now includes the apphost version and architecture, just like we do if there's no runtime at all.

Also some PR feedback fixes.

vitek-karas commented:
I improved the dialog for the case of old runtime, it now looks very similar to the case where there's no runtime at all. Most notably it contains apphost version and architecture:
![image](https://user-images.githubusercontent.com/10670590/164675056-40126959-bb7e-49c5-902e-7ac4f87cdaf1.png)
